nsqd/channel.go

```diff
@@ -325,6 +325,8 @@ func (c *Channel) put(m *Message) error {
 				c.name, err)
 			c.ctx.nsqd.SetHealth(err)
 			return err
+		} else if !c.ctx.nsqd.IsHealthy() {
+			c.ctx.nsqd.SetHealth(nil)
```

Comment:
I think the simplest approach is to always `c.ctx.nsqd.SetHealth(err)` right after line 322 (before the `err` check).  As you mentioned in the description, you didn't do this so that we didn't add a mutex to this code path.

I propose that we instead change the way health is stored so that it is just an `error` (and we set it via [`atomic.Value`](https://golang.org/pkg/sync/atomic/#Value)).  This eliminates the mutex and would allow us to do make the above change.

What do you think?